src/drivers/net/mcst/l_e1000.c

```diff
@@ -0,0 +1,728 @@
+	memcpy(&dev->dev_addr[0], addr, sizeof(dev->dev_addr));
+	log_debug("mac addr: %x:%x:%x:%x:%x:%x\n", dev->dev_addr[0], dev->dev_addr[1], dev->dev_addr[2], dev->dev_addr[3], dev->dev_addr[4], dev->dev_addr[5]);
+//	l_e1000_open(dev);
+	return 0;
+}
+
+static int l_e1000_stop(struct net_device *dev) {
+	return 0;
+}
+
+static const struct net_driver l_e1000_drv_ops = {
+	.xmit = l_e1000_xmit,
+	.start = l_e1000_open,
+	.stop = l_e1000_stop,
+	.set_macaddr = l_e1000_set_mac_address
+};
+
+static int l_e1000_rx(struct l_e1000_priv *ep) {
+	uint8_t * buf = NULL;
+	int msg_len;
+	int entry;
+
+	entry  = ep->cur_rx & RX_RING_MOD_MASK;
+
+	/* If we own the next entry, it's a new packet. Send it up. */
+	while ((le16toh(ep->rx_ring[entry].status)) >= 0) {
+		struct sk_buff *skb;
+		int status;
+
+		ep->cur_rx++;
+
+		status = le16toh(ep->rx_ring[entry].status);
+		skb = ep->rx_skbs[entry];
+
+		if ((status & 0xff00) != (RD_ENP|RD_STP)) {/* There was an error. */
+			msg_len = 0;
+			if (status & RD_ENP) {
+				log_debug(" ENP");
+			} /* No detailed rx_errors counter to increment at the */
+			/* end of a packet.*/
+			if (status & RD_FRAM) {
+				log_debug(" FRAM");
+			}
+			if (status & RD_OFLO) {
+				log_debug(" OFLO ");
+			}
+	    	if (status & RD_CRC) {
+				log_debug(" CRC ");
+			}
+			if (status & RD_BUFF) {
+				log_debug(" BUFF ");
+			}
+			skb_free(skb);
+		} else {
+			buf = (void *)(uintptr_t)le32toh(ep->rx_ring[entry].base);
+			msg_len =  (le16toh(ep->rx_ring[entry].msg_length) & 0xfff) - ETH_FCS_LEN;
+			show_packet(buf, msg_len, " --- BUF received packet print:");
+			skb->dev = ep->netdev;
+			skb->len = msg_len;
+
+			netif_rx(skb);
+		}
+
+		ep->rx_skbs[entry] = l_e1000_new_rx_buff(ep, entry);
+
+		entry  = ep->cur_rx & RX_RING_MOD_MASK;
+	}
+	return 0;
+}
+
+static irq_return_t l_e1000_interrupt(unsigned int irq_num, void *dev_id) {
+	irq_return_t ret = IRQ_NONE;
+	struct net_device *dev = dev_id;
+	struct l_e1000_priv *ep;
+	uint16_t csr0;//, csr_ack;
+
+	ep = netdev_priv(dev);
+	csr0 = e1000_read_e_csr(ep);
+	if (!(csr0 & INTR)) {
+		return IRQ_NONE; /* Not our interrupt */
+	}
+
+	log_debug("irq csr0(%x)", csr0);
+
+	csr0 &= (BABL|CERR|MISS|MERR|RINT|TINT);
+	e1000_write_e_csr(ep, csr0|IDON);
+
+	if (csr0 & TINT) {
+		int status = le16toh(ep->tx_ring->status);
+		if (status & TD_ERR) {
+			int err_status = le32toh(ep->tx_ring->misc);
+			log_error("Tx error status=%04x err_status=%08x", status, err_status);
+		}
+		ep->tx_ring->status = 0;
+	}
+
+	if (csr0 & MERR) {
+		log_debug("irq MERR");
+		csr0 &= ~MERR;
+	}
+	if (csr0 & BABL) {
+		log_debug("irq BABL");
+		csr0 &= ~BABL;
+	}
+
+	if (csr0 & MISS) {
+		//ep->rx_skipped++;
+		csr0 &= ~MISS;
+		/* The device did't set RINT in case of buffers are full
+		* and MISS is set. So we do this manually.
+		*/
+		csr0 |= RINT;
+		log_debug("irq BABL");
+	}
+
+	if (csr0 & RINT) {
+		l_e1000_rx(ep);
+	}
+
+	e1000_write_e_csr(ep, INEA);
+
+	ret = IRQ_HANDLED;
+	return ret;
```

Comment:
Can we just `return IRQ_HANDLED` and remove `ret` variable since it is not used anymore?

Reply:
done